File: appkit-creating-and-customizing-the-touch-bar/Objective-C/NSTouchBar Catalog/TestViewControllers/StepperViewController.m
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
View controller responsible for showing NSStepperTouchBarItem in an NSTouchBar instance.
*/

#import "StepperViewController.h"

static NSTouchBarCustomizationIdentifier StepperItemCustomizationIdentifier = @"com.TouchBarCatalog.stepperItemController";
static NSTouchBarItemIdentifier StepperItemIdentifier = @"com.TouchBarCatalog.stepperItem";

@interface StepperViewController () <NSTouchBarDelegate>

@property (weak) IBOutlet NSButton *useCustomDraw;

@end

#pragma mark -

@implementation StepperViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
}

// The user chose Custom Button Color checkbox to change the button colors.
- (IBAction)customize:(id)sender
{
    // Set to nil so you can call makeTouchBar again to recreate the NSTouchBar instance.
    self.touchBar = nil;
}

#pragma mark NSTouchBarProvider

- (NSTouchBar *)makeTouchBar
{
    NSTouchBar *bar = [[NSTouchBar alloc] init];
    bar.delegate = self;
    
    bar.customizationIdentifier = StepperItemCustomizationIdentifier;
    
    // Set the default ordering of items.
    bar.defaultItemIdentifiers = @[StepperItemIdentifier, NSTouchBarItemIdentifierOtherItemsProxy];
    
    bar.customizationAllowedItemIdentifiers = @[StepperItemIdentifier];
    
    return bar;
}

#pragma mark NSTouchBarDelegate

// The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
- (nullable NSTouchBarItem *)touchBar:(NSTouchBar *)touchBar makeItemForIdentifier:(NSTouchBarItemIdentifier)identifier
{
    if ([identifier isEqualToString:StepperItemIdentifier])
    {
        NSStepperTouchBarItem *stepperItem = nil;
        
        if (self.useCustomDraw.state == NSControlStateValueOff) {
            NSNumberFormatter *numberFormatter = [[NSNumberFormatter alloc] init];
            numberFormatter.numberStyle = NSNumberFormatterPercentStyle;
            numberFormatter.maximumFractionDigits = 0;
            numberFormatter.multiplier = @1;
            
            stepperItem = [NSStepperTouchBarItem stepperTouchBarItemWithIdentifier:identifier formatter: numberFormatter];
        }
        else
        {
            stepperItem = [NSStepperTouchBarItem stepperTouchBarItemWithIdentifier:identifier drawingHandler:^void(NSRect rect, double value) {
                [NSGraphicsContext saveGraphicsState];

                // Draw the stepper value.
                NSString *valueStr = [NSString stringWithFormat:@"%.0f", value];
                NSFont *font = [NSFont systemFontOfSize:12.0];
                
                NSDictionary *attributes = [NSMutableDictionary dictionaryWithObjectsAndKeys:
                                            [NSColor whiteColor], NSForegroundColorAttributeName,
                                            font, NSFontAttributeName,
                                            nil];
                NSAttributedString *attrString = [[NSAttributedString alloc] initWithString:valueStr attributes:attributes];

                NSSize valueSize = NSMakeSize(CGFLOAT_MAX, CGFLOAT_MAX);
                NSRect valueRect = [attrString boundingRectWithSize:valueSize options:NSStringDrawingUsesFontLeading | NSStringDrawingUsesLineFragmentOrigin];
                NSPoint pt = NSMakePoint((rect.size.width - valueRect.size.width) / 2, (rect.size.height - valueRect.size.height) / 2);
                [attrString drawAtPoint: pt];
         
                // Adorn the stepper value.
                NSBezierPath *path = [NSBezierPath bezierPathWithRect:rect];
                [[NSColor whiteColor] set];
                [path stroke];
                
                [NSGraphicsContext restoreGraphicsState];
            }];
        }
           
        stepperItem.maxValue = 100;
        stepperItem.minValue = 1;
        stepperItem.increment = 10;
        stepperItem.value = 50;
        
        return stepperItem;
    }
    
    return nil;
}

@end

File: appkit-creating-and-customizing-the-touch-bar/Swift/NSTouchBar Catalog/StepperViewController.swift
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
View controller responsible for showing NSStepperTouchBarItem in an NSTouchBar instance.
*/

import Cocoa

class StepperViewController: NSViewController {
    let stepper = NSTouchBarItem.Identifier("com.TouchBarCatalog.TouchBarItem.stepper")
    let stepperBar = NSTouchBar.CustomizationIdentifier("com.TouchBarCatalog.stepperBar")
    
    @IBOutlet weak var useCustomDraw: NSButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func makeTouchBar() -> NSTouchBar? {
        let touchBar = NSTouchBar()
        touchBar.delegate = self
        touchBar.customizationIdentifier = stepperBar
        touchBar.defaultItemIdentifiers = [stepper]
        return touchBar
    }
    
    // MARK: - Action Functions
    
    @IBAction func customize(_ sender: AnyObject) {
        // This creates a call to makeTouchBar.
        touchBar = nil
    }

}

// MARK: - NSTouchBarDelegate

extension StepperViewController: NSTouchBarDelegate {

    // The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
    func touchBar(_ touchBar: NSTouchBar, makeItemForIdentifier identifier: NSTouchBarItem.Identifier) -> NSTouchBarItem? {
        switch identifier {
        case stepper:
            var stepperItem: NSStepperTouchBarItem
            
            if useCustomDraw.state == NSControl.StateValue.on {
                // Create the stepper touch bar item using a custom drawing handler.
                stepperItem = NSStepperTouchBarItem(identifier: identifier, drawingHandler: { frame, value in
                    NSGraphicsContext.saveGraphicsState()

                    // Draw the stepper value.
                    let valueStr = String(format: "%.0f", value)
                    let font = NSFont.systemFont(ofSize: 12)
                    let string = NSAttributedString(string: valueStr, attributes:
                                                        [ NSAttributedString.Key.foregroundColor: NSColor.white,
                                                          NSAttributedString.Key.font: font])
                    let valueSize = NSSize(width: CGFloat.greatestFiniteMagnitude, height: CGFloat.greatestFiniteMagnitude)
                    let valueRect = string.boundingRect(with: valueSize, options: [.usesFontLeading, .usesLineFragmentOrigin])
                    string.draw(at: CGPoint(x: (frame.width - valueRect.width) / 2, y: (frame.height - valueRect.height) / 2))

                    // Adorn the stepper value.
                    let fillColor = NSColor.white
                    let path = NSBezierPath(rect: frame)
                    fillColor.set()
                    path.stroke()
                    
                    NSGraphicsContext.restoreGraphicsState()
                })
            } else {
                // Create the stepper touch bar item with built-in value drawing, using a number formatter.
                let formatter = NumberFormatter()
                formatter.numberStyle = .percent
                formatter.maximumFractionDigits = 0
                formatter.multiplier = 1
                
                stepperItem = NSStepperTouchBarItem(identifier: identifier, formatter: formatter)
            }
            
            stepperItem.maxValue = 100
            stepperItem.minValue = 1
            stepperItem.increment = 10
            stepperItem.value = 50
            
            return stepperItem
        default:
            return nil
        }
    }

}
